icu4c/source/test/intltest/rbbitst.cpp

```diff
@@ -4098,53 +4287,72 @@ void RBBITest::RunMonkey(BreakIterator *bi, RBBIMonkeyKind &mk, const char *name
                     }
                 }
 
-                // Format looks like   "<data>\\\uabcd\uabcd\\\U0001abcd...</data>"
-                UnicodeString errorText = "<data>";
-                /***if (strcmp(errorType, "next()") == 0) {
-                    startContext = 0;
-                    endContext = testText.length();
+                // Formatting of each line includes:
+                //   character code
+                //   reference break: '|' -> a break, '.' -> no break
+                //   actual break:    '|' -> a break, '.' -> no break
+                //   (name of character clase)
+                //   Unicode name of character
+                //   '-->' indicates location of the difference.
 
-                    printStringBreaks(testText, expected, expectedCount);
-                }***/
+                UErrorCode status = U_ZERO_ERROR;
+                const char *badLocale = bi->getLocaleID(ULOC_ACTUAL_LOCALE, status);
+                errln("\n%s break monkey test error [%s].  %s. Operation = %s; Random seed = %d;  buf Idx = %d",
+                    name, badLocale, (expectedBreaks[i]? "break expected but not found" : "break found but not expected"),
+                    errorType, seed, i);
 
-                for (ci=startContext; ci<endContext;) {
-                    UnicodeString hexChars("0123456789abcdef");
+                for (ci=startContext; ci<=endContext;) {
+                    UnicodeString errorText = "";
                     UChar32  c;
-                    int      bn;
                     c = testText.char32At(ci);
+                    
+                    char buffer[200];
+                    std::string format = "    %3i : ";
                     if (ci == i) {
-                        // This is the location of the error.
-                        errorText.append("<?>");
-                    } else if (expectedBreaks[ci] != 0) {
-                        // This a non-error expected break position.
-                        errorText.append("\\");
+                        format =  "--> %3i : ";  // Error position
                     }
-                    if (c < 0x10000) {
-                        errorText.append("\\u");
-                        for (bn=12; bn>=0; bn-=4) {
-                            errorText.append(hexChars.charAt((c>>bn)&0xf));
-                        }
-                    } else {
-                        errorText.append("\\U");
-                        for (bn=28; bn>=0; bn-=4) {
-                            errorText.append(hexChars.charAt((c>>bn)&0xf));
-                        }
+                    sprintf(buffer, format.c_str(), ci);
+                    errorText.append(buffer);
```

Comment:
"buffer" is being changed to a temporary UnicodeString via the default constructor, which drags in code page conversion, and is non-obvious in what it's doing.  (there is no UnicodeString.append(const char *) function).

At a minimum, be explicit about it.
Optional: think about whether the error message assembly can be done purely in either 8 or 16 bit space, avoiding conversions back and forth. It's not clear that it would be simpler overall, but worth thinking about.

Reply:
This is now restructured to remove errorText, using only a char array.